File: frontend/src/worker/routes/transactions.ts
import { Hono } from 'hono'
import { requireCsrf, requireSession } from '../middleware/auth'
import { TransactionRepository } from '../repositories/transactionRepository'
import { manualTransactionSchema, transactionUpdateSchema } from '../schemas'
import { TransactionService } from '../services/transactionService'
import type { AppEnvironment } from '../types'
import { HttpError } from '../utils/http'
import { normalizeDateTime } from '../utils/time'
import { numberQuery, parseBody } from './helpers'

const SORT_VALUES = new Set(['newest', 'oldest', 'amount_high', 'amount_low'])

export const transactionRoutes = new Hono<AppEnvironment>()
transactionRoutes.use('/transactions', requireSession)
transactionRoutes.use('/transactions/*', requireSession)

transactionRoutes.get('/transactions', async (c) => {
  const query = c.req.query()
  const page = numberQuery(query.page, 1, 1, 1_000_000)
  const pageSize = numberQuery(query.page_size, 50, 1, 200)
  const sort = query.sort ?? 'newest'
  if (!SORT_VALUES.has(sort)) throw new HttpError(422, 'Invalid transaction sort')
  const minimumMinor = query.min_amount_minor === undefined ? undefined : numberQuery(query.min_amount_minor, 0, 0, Number.MAX_SAFE_INTEGER)
  const maximumMinor = query.max_amount_minor === undefined ? undefined : numberQuery(query.max_amount_minor, 0, 0, Number.MAX_SAFE_INTEGER)
  const result = await new TransactionRepository(c.env.DB).list({
    page,
    pageSize,
    search: query.search?.trim() || undefined,
    dateFrom: query.date_from ? normalizeDateTime(query.date_from) : undefined,
    dateTo: query.date_to ? normalizeDateTime(query.date_to) : undefined,
    type: query.type,
    categoryId: query.category_id,
    paymentMethodId: query.payment_method_id,
    source: query.source,
    reviewStatus: query.review_status,
    currency: query.currency?.toUpperCase(),
    minimumMinor,
    maximumMinor,
    sort: sort as 'newest' | 'oldest' | 'amount_high' | 'amount_low',
  })
  return c.json({ ...result, page, page_size: pageSize })
})

transactionRoutes.post('/transactions', requireCsrf, async (c) => {
  const payload = await parseBody(c, manualTransactionSchema)
  const result = await new TransactionService(c.env.DB).ingestManual(payload)
  return c.json({
    success: true,
    result: result.result,
    transaction_id: result.transaction.id,
    category: result.transaction.category_name,
    review_status: result.transaction.review_status,
    duplicate: result.duplicate,
  }, 201)
})

transactionRoutes.get('/transactions/:id', async (c) => c.json(await new TransactionService(c.env.DB).get(c.req.param('id'))))

transactionRoutes.patch('/transactions/:id', requireCsrf, async (c) => {
  const payload = await parseBody(c, transactionUpdateSchema)
  return c.json(await new TransactionService(c.env.DB).update(c.req.param('id'), payload))
})

transactionRoutes.delete('/transactions/:id', requireCsrf, async (c) => {
  await new TransactionService(c.env.DB).delete(c.req.param('id'))
  return c.body(null, 204)
})

File: backend/app/api/transactions.py
from __future__ import annotations

from datetime import datetime

from fastapi import APIRouter, Depends, Query, Response, status
from sqlalchemy import func, or_, select
from sqlalchemy.orm import Session, selectinload

from app.api.dependencies import get_transaction_or_404
from app.api.presenters import ingestion_response, transaction_response
from app.auth.dependencies import AuthenticatedSession, get_authenticated_session, require_csrf
from app.database.session import get_db
from app.models import LedgerTransaction
from app.models.enums import ReviewStatus, TransactionSource, TransactionType
from app.schemas.transaction import (
    IngestionResponse,
    ManualTransactionRequest,
    TransactionListResponse,
    TransactionResponse,
    TransactionUpdate,
)
from app.services.ingestion import ingest_manual
from app.services.merchant_normalization import normalize_merchant
from app.services.money import to_minor_units


router = APIRouter(prefix="/api/v1/transactions", tags=["transactions"])


def _detail_query():
    return select(LedgerTransaction).options(
        selectinload(LedgerTransaction.category),
        selectinload(LedgerTransaction.payment_method),
    )


@router.get("", response_model=TransactionListResponse)
def list_transactions(
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=50, ge=1, le=200),
    search: str | None = None,
    date_from: datetime | None = None,
    date_to: datetime | None = None,
    transaction_type: TransactionType | None = Query(default=None, alias="type"),
    category_id: str | None = None,
    payment_method_id: str | None = None,
    source: TransactionSource | None = None,
    review_status: ReviewStatus | None = None,
    currency: str | None = None,
    min_amount_minor: int | None = None,
    max_amount_minor: int | None = None,
    sort: str = Query(default="newest", pattern="^(newest|oldest|amount_high|amount_low)$"),
    _authenticated: AuthenticatedSession = Depends(get_authenticated_session),
    database: Session = Depends(get_db),
) -> TransactionListResponse:
    filters = []
    if search:
        needle = f"%{search.strip()}%"
        filters.append(
            or_(
                LedgerTransaction.merchant_raw.ilike(needle),
                LedgerTransaction.merchant_normalized.ilike(needle),
                LedgerTransaction.note.ilike(needle),
                LedgerTransaction.purpose.ilike(needle),
            )
        )
    if date_from:
        filters.append(LedgerTransaction.transaction_date >= date_from)
    if date_to:
        filters.append(LedgerTransaction.transaction_date <= date_to)
    if transaction_type:
        filters.append(LedgerTransaction.type == transaction_type)
    if category_id:
        filters.append(LedgerTransaction.category_id == category_id)
    if payment_method_id:
        filters.append(LedgerTransaction.payment_method_id == payment_method_id)
    if source:
        filters.append(LedgerTransaction.source == source)
    if review_status:
        filters.append(LedgerTransaction.review_status == review_status)
    if currency:
        filters.append(LedgerTransaction.currency_code == currency.upper())
    if min_amount_minor is not None:
        filters.append(LedgerTransaction.amount_minor >= min_amount_minor)
    if max_amount_minor is not None:
        filters.append(LedgerTransaction.amount_minor <= max_amount_minor)

    order_by = {
        "newest": LedgerTransaction.transaction_date.desc(),
        "oldest": LedgerTransaction.transaction_date.asc(),
        "amount_high": LedgerTransaction.amount_minor.desc(),
        "amount_low": LedgerTransaction.amount_minor.asc(),
    }[sort]
    total = database.scalar(
        select(func.count()).select_from(LedgerTransaction).where(*filters)
    ) or 0
    items = database.scalars(
        _detail_query()
        .where(*filters)
        .order_by(order_by, LedgerTransaction.id.desc())
        .offset((page - 1) * page_size)
        .limit(page_size)
    ).all()
    return TransactionListResponse(
        items=[transaction_response(item) for item in items],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.post("", response_model=IngestionResponse, status_code=201)
def create_transaction(
    payload: ManualTransactionRequest,
    _authenticated: AuthenticatedSession = Depends(require_csrf),
    database: Session = Depends(get_db),
) -> IngestionResponse:
    return ingestion_response(ingest_manual(database, payload))


@router.get("/{transaction_id}", response_model=TransactionResponse)
def get_transaction(
    transaction_id: str,
    _authenticated: AuthenticatedSession = Depends(get_authenticated_session),
    database: Session = Depends(get_db),
) -> TransactionResponse:
    transaction = database.scalar(
        _detail_query().where(LedgerTransaction.id == transaction_id)
    )
    if transaction is None:
        get_transaction_or_404(database, transaction_id)
    return transaction_response(transaction)


@router.patch("/{transaction_id}", response_model=TransactionResponse)
def update_transaction(
    transaction_id: str,
    payload: TransactionUpdate,
    _authenticated: AuthenticatedSession = Depends(require_csrf),
    database: Session = Depends(get_db),
) -> TransactionResponse:
    transaction = get_transaction_or_404(database, transaction_id)
    changes = payload.model_dump(exclude_unset=True)
    amount = changes.pop("amount", None)
    currency = changes.pop("currency", None)
    merchant = changes.pop("merchant", None)
    if currency is not None:
        transaction.currency_code = currency
    if amount is not None:
        transaction.amount_minor = to_minor_units(amount, transaction.currency_code)
    if merchant is not None:
        transaction.merchant_raw = merchant
        transaction.merchant_normalized = normalize_merchant(merchant)
    for field, value in changes.items():
        setattr(transaction, field, value)
    database.commit()
    transaction = database.scalar(
        _detail_query().where(LedgerTransaction.id == transaction_id)
    )
    return transaction_response(transaction)


@router.delete("/{transaction_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_transaction(
    transaction_id: str,
    _authenticated: AuthenticatedSession = Depends(require_csrf),
    database: Session = Depends(get_db),
) -> Response:
    transaction = get_transaction_or_404(database, transaction_id)
    database.delete(transaction)
    database.commit()
    return Response(status_code=status.HTTP_204_NO_CONTENT)
